### PR TITLE
Add newline for bracket-less arrow functions that return calls

### LINKED ISSUE #797: Missing \n in {}-less arrow function calls
```js
const testResults = results.testResults.map(testResult =>
  formatResult(testResult, formatter, reporter)
);
```

gets formatted into

```js
const testResults = results.testResults.map(testResult =>
  formatResult(testResult, formatter, reporter));
```

We should probably keep the `)` where it was for symmetry.

https://prettier.github.io/prettier/#%7B%22content%22%3A%22const%20testResults%20%3D%20results.testResults.map(testResult%20%3D%3E%5Cn%20%20formatResult(testResult%2C%20formatter%2C%20reporter)%5Cn)%3B%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

### LINKED ISSUE #922: Odd indentation in expect call.
See: https://github.com/cpojer/jest/blob/master/packages/jest-mock/src/__tests__/jest-mock-test.js#L166

before:
```js
    it('mocks regexp instances', () => {
      expect(
        () => moduleMocker.generateFromMetadata(moduleMocker.getMetadata(/a/)),
      ).not.toThrow();
    });
```

after:
```js
    it('mocks regexp instances', () => {
      expect(() =>
        moduleMocker.generateFromMetadata(
          moduleMocker.getMetadata(/a/),
        )).not.toThrow();
    });
```

I would at least expect:

```js
    it('mocks regexp instances', () => {
      expect(() =>
        moduleMocker.generateFromMetadata(
          moduleMocker.getMetadata(/a/),
        ),
      ).not.toThrow();
    });
```

but the current behavior seems wrong.

https://prettier.github.io/prettier/#%7B%22content%22%3A%22it('mocks%20regexp%20instances'%2C%20()%20%3D%3E%20%7B%5Cn%20%20expect(%5Cn%20%20%20%20()%20%3D%3E%20moduleMocker.generateFromMetadata(moduleMocker.getMetadata(%2Fa%2F))%2C%5Cn%20%20).not.toThrow()%3B%5Cn%7D)%3B%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22%22%2C%22doc%22%3Afalse%7D%7D

### LINKED ISSUE #974: Edge-cases with `expect` + callback as first param
I found some "edge cases" when playing around with the results I got when running Prettier on my tests.

```js
expect(() => asyncRequest({ url: "/test-endpoint" }))
  .toThrowError(/Required parameter/);

expect(() => asyncRequest({ url: "/test-endpoint-but-with-a-long-url" }))
  .toThrowError(/Required parameter/);

expect(() => asyncRequest({ url: "/test-endpoint-but-with-a-suuuuuuuuper-long-url" }))
  .toThrowError(/Required parameter/);

expect(() => asyncRequest({ type: "foo", url: "/test-endpoint" }))
  .not.toThrowError();

expect(() => asyncRequest({ type: "foo", url: "/test-endpoint-but-with-a-long-url" }))
  .not.toThrowError();
```

transforms into:

```js
expect(() => asyncRequest({ url: "/test-endpoint" })).toThrowError(
  /Required parameter/
);

expect(() =>
  asyncRequest({ url: "/test-endpoint-but-with-a-long-url" })).toThrowError(
  /Required parameter/
);

expect(() =>
  asyncRequest({
    url: "/test-endpoint-but-with-a-suuuuuuuuper-long-url"
  })).toThrowError(/Required parameter/);

expect(() =>
  asyncRequest({ type: "foo", url: "/test-endpoint" })).not.toThrowError();

expect(() =>
  asyncRequest({
    type: "foo",
    url: "/test-endpoint-but-with-a-long-url"
  })).not.toThrowError();
```

Number 2 feels like a bug to me, the others might be "as intended" (just different from what I expected).

For me, in the last three it kinda feels like the assertions are _within_ the callback (as we're not putting the `)` on its own line it's "difficult" to see we're doing something to the result). I'm not sure what result I think would be best here (and it might just be "it's different so it's just about getting used to it"), but to have some examples for discussion these seem easier to read:

```js
expect(() => asyncRequest({ url: "/test-endpoint" }))
  .toThrowError(/Required parameter/);

expect(() => asyncRequest({ url: "/test-endpoint-but-with-a-long-url" }))
  .toThrowError(/Required parameter/);

expect(() =>
  asyncRequest({ url: "/test-endpoint-but-with-a-suuuuuuuuper-long-url" })
).toThrowError(/Required parameter/);

expect(() => asyncRequest({ type: "foo", url: "/test-endpoint" }))
  .not.toThrowError();

expect(() =>
  asyncRequest({
    type: "foo",
    url: "/test-endpoint-but-with-a-long-url"
  })
).not.toThrowError();
```

Or maybe:

```js
expect(() =>
  asyncRequest({ url: "/test-endpoint-but-with-a-suuuuuuuuper-long-url " })
)
  .toThrowError(/Required parameter/);

expect(() =>
  asyncRequest({
    type: "foo",
    url: "/test-endpoint-but-with-a-long-url"
  })
)
  .not.toThrowError();
```

But this doesn't look good:

```
expect(() =>
  asyncRequest({ url: "/test-endpoint-but-with-a-suuuuuuuuper-long-url " }))
  .toThrowError(/Required parameter/);
```

So all in all: not sure, just thought I'd open this to hear your thoughts

https://prettier.github.io/prettier/#%7B%22content%22%3A%22expect(()%20%3D%3E%20asyncRequest(%7B%20url%3A%20%60%2Ftest-endpoint%60%20%7D))%5Cn%20%20.toThrowError(%2FRequired%20parameter%2F)%3B%5Cn%5Cnexpect(()%20%3D%3E%20asyncRequest(%7B%20url%3A%20%60%2Ftest-endpoint-but-with-a-long-url%60%20%7D))%5Cn%20%20.toThrowError(%2FRequired%20parameter%2F)%3B%5Cn%5Cnexpect(()%20%3D%3E%20asyncRequest(%7B%20url%3A%20%60%2Ftest-endpoint-but-with-a-suuuuuuuuper-long-url%60%20%7D))%5Cn%20%20.toThrowError(%2FRequired%20parameter%2F)%3B%5Cn%5Cnexpect(()%20%3D%3E%20asyncRequest(%7B%20type%3A%20%60foo%60%2C%20url%3A%20%60%2Ftest-endpoint%60%20%7D))%5Cn%20%20.not.toThrowError()%3B%5Cn%5Cnexpect(()%20%3D%3E%20asyncRequest(%7B%20type%3A%20%60foo%60%2C%20url%3A%20%60%2Ftest-endpoint-but-with-a-long-url%60%20%7D))%5Cn%20%20.not.toThrowError()%3B%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22doc%22%3Afalse%7D%7D

### PR DESCRIPTION
**Pull request by vjeux**
I've tried a lot of places where to fix this and this is the only one that gives correct results. This is likely not exhaustive but works for that use case. It's been reported twice in issues and I've seen it happen a few other times so we probably want to get it fixed.

Fixes #922
Fixes #797
Fixes #974

### CODE REVIEW

**Comment by cpojer** [comment]
❤️

**Comment by kjbekkelund** [comment]
I ❤️ this change!

File: tests/method-chain/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -45,14 +45,16 @@ export default store => {
...
+      )
+    ,
     error =>
       next(
         actionWith({
           type: failureType,
           error: error.message || \\"Something bad happened\\"
         })
       )
+
```
Comment (vjeux) [inline tests/method-chain/__snapshots__/jsfmt.spec.js.snap]: All of the other examples are better but this one output is very weird.

**Comment by vjeux** [comment]
@difelice great catch! I can reproduce, looking into a solution now.

For the docs that do not build, I fixed it in https://github.com/prettier/prettier/pull/1024, sorry about that.

**Comment by vjeux** [comment]
I just pushed a fix for it, here is the snapshot that changed:

image

I forgot to add ifBreak and "all" option /facepalm

**Comment by difelice** [comment]
Thank-you!

**Comment by difelice** [comment]
@vjeux I found another issue:

Add this case in `arrow_call.js`:
```js
func(
  veryLoooooooooooooooooooooooongName,
  veryLooooooooooooooooooooooooongName =>
    veryLoooooooooooooooongName.something()
);
```

An extra indented line is added (with a comma using `{ trailingComma: 'all' }`):
image
image

I hope this helps to reproduce.

I just read `commands.md`, I'll try to give a hand.

Thanks.

**Comment by vjeux** [comment]
Excellent, thanks for digging through it. So it double prints when the outer breaks. That's going to be interesting to fix!

**Comment by real34** [comment]
Hi! Thanks for this work, along with #1066.

If it can help to gain confidence with these improvements (and maybe add a few edge cases) I created [a compilation of related issues](https://prettier.github.io/prettier/#%7B%22content%22%3A%22const%20a%20%3D%20Observable%5Cn%20%20.fromPromise(axiosInstance.post('%2Fcarts%2Fmine'))%5Cn%20%20.map((response)%20%3D%3E%20response.data)%5Cn%5Cnconst%20b%20%3D%20Observable.fromPromise(axiosInstance.get(url))%5Cn%20%20.map((response)%20%3D%3E%20response.data)%5Cn%5Cn%5Cnconst%20CategoryLoader%20%3D%20(axiosInstance)%20%3D%3E%20%7B%5Cn%20%20const%20load%20%3D%20(id)%20%3D%3E%20(%5Cn%20%20%5CtObservable%5Cn%20%20%20%20%20%20.fromPromise(axiosInstance.get(%60%2Fcategories%2F%24%7Bid%7D%60))%5Cn%20%20%20%20%20%20.map((response)%20%3D%3E%20response.data)%5Cn%20%20%20%20%20%20.map((category)%20%3D%3E%20(%7B%5Cn%20%20%20%20%20%20%20%20...convertCategoryMainAttributesForFront(category)%2C%5Cn%20%20%20%20%20%20%20%20children%3A%20category.children.split('%2C').filter((id)%20%3D%3E%20id%20!%3D%3D%20'')%5Cn%20%20%20%20%20%20%7D))%5Cn%20%20)%5Cn%7D%5Cn%5Cnconst%20loadPricesBatch%20%3D%20(skus)%20%3D%3E%20%7B%5Cn%20%20const%20params%20%3D%20%7B%7D%5Cn%20%20return%20Observable%5Cn%20%20%20%20.fromPromise(axiosInstance.get('%2Ffrontcommerce%2Fprice'%2C%20%7Bparams%7D))%5Cn%20%20%20%20.map((response)%20%3D%3E%20response.data)%5Cn%20%20%20%20.map(reorderForSkus(skus))%5Cn%7D%5Cn%5Cnconst%20changePassword%20%3D%20(currentPassword%2C%20newPassword)%20%3D%3E%20%7B%5Cn%20%20return%20Observable.fromPromise(axiosInstance.put('%2Fcustomers%2Fme%2Fpassword'%2C%20%7B%20currentPassword%2C%20newPassword%20%7D))%5Cn%20%20%20%20.map((response)%20%3D%3E%20(%7B%20success%3A%20response.data%20%7D))%5Cn%7D%5Cn%5Cnconst%20optionIsPartOf%20%3D%20(configuration)%20%3D%3E%20(option)%20%3D%3E%20(%5Cn%20%20configuration.some((%7Bid%2C%20value%7D)%20%3D%3E%20id%20%3D%3D%3D%20option.id%20%26%26%20value%20%3D%3D%3D%20option.value)%5Cn)%5Cn%5Cn%2F%2F%20splitting%20the%20callback%20params%20on%20a%20newline%20as%20below%5Cn%2F%2F%20seems%20much%20readable%20imo%20than%20the%20whole%20%60new%20Dataloader(...)%60%20%5Cnconst%20makeUrlLoader%20%3D%20(type)%20%3D%3E%20new%20DataLoader(%5Cn%20%20(ids)%20%3D%3E%20loadUrlsBatch(type%2C%20ids).toPromise()%5Cn)%5Cn%5Cn%5Cnconst%20reorderForSkus%20%3D%20(skus)%20%3D%3E%20(data)%20%3D%3E%20(%5Cn%20%20skus.map((sku)%20%3D%3E%20data.find(%5Cn%20%20%20%20(result)%20%3D%3E%20result.sku%20%3D%3D%3D%20sku%5Cn%20%20)%20%7C%7C%20null)%5Cn)%5Cn%5Cnconst%20reorderForIds%20%3D%20(ids%2C%20idKey)%20%3D%3E%20(data)%20%3D%3E%20%7B%5Cn%20%20return%20ids.map((id)%20%3D%3E%20data.find(%5Cn%20%20%20%20(result)%20%3D%3E%20result%5BidKey%5D.toString()%20%3D%3D%3D%20id.toString()%5Cn%20%20)%20%7C%7C%20null)%5Cn%7D%5Cn%5Cnconst%20apolloMutationToRxObservable%20%3D%20(apolloPromise)%20%3D%3E%20%7B%5Cn%20%20return%20Observable.create((observer)%20%3D%3E%20%7B%5Cn%20%20%20%20%2F%2F%20the%20then%20callback%20content%2C%20when%20pretified%20as%20idented%5Cn%20%20%20%20%2F%2F%20would%20be%20better%20wrapped%20with%20parenthesis%20imo%20%5Cn%20%20%20%20apolloPromise%5Cn%20%20%20%20%20%20.then((response)%20%3D%3E%20observer.next(%7Bstatus%3A%20'success'%2C%20response%3A%20response%7D))%5Cn%20%20%20%20%20%20.catch((error)%20%3D%3E%20%7B%5Cn%20%20%20%20%20%20%20%20let%20response%20%3D%20error%5Cn%20%20%20%20%20%20%20%20observer.next(%7Bstatus%3A%20'error'%2C%20response%7D)%5Cn%20%20%20%20%20%20%7D)%5Cn%20%20%7D)%5Cn%7D%5Cn%5Cn%20%5CnwithHandlers(%7B%5Cn%20%20onImageClick%3A%20(props)%20%3D%3E%20(image)%20%3D%3E%20%7B%5Cn%5Ct%2F%2F%20foo%20%20%20%20%5Cn%20%20%7D%5Cn%7D)%5Cn%5Cnconst%20EnhanceDelivery%20%3D%20(fragments)%20%3D%3E%20compose(%5Cn%20%20makeCommandDispatcherWithMessageFeedback(%7B%5Cn%20%20%20%20onSubmit%3A%20(props)%20%3D%3E%20(e)%20%3D%3E%20%7B%5Cn%20%20%20%20%20%20e.preventDefault()%5Cn%20%20%20%20%20%20%2F%2F%20...%5Cn%20%20%20%20%7D%5Cn%20%20%7D)%5Cn)%5Cn%5Cnconst%20hasPrice%20%3D%20(config)%20%3D%3E%20config%20%26%26%20config.product%20%26%26%20typeof%20config.product.price%20!%3D%3D%20'undefined'%5Cnconst%20priceOf%20%3D%20(config)%20%3D%3E%20config.product.price%5Cnconst%20cheapestFirst%20%3D%20(a%2C%20b)%20%3D%3E%20priceOf(b)%20-%20priceOf(a)%5Cn%5Cnconst%20logFailedActions%20%3D%20(store)%20%3D%3E%20(next)%20%3D%3E%20(action)%20%3D%3E%20%7B%5Cn%20%20%20%2F%2F%20...%5Cn%7D%5Cn%5Cnconst%20cart%20%3D%20%7B%5Cn%20%20type%3A%20CartType%2C%5Cn%20%20resolve%3A%20(_%2C%20args%2C%20%7B%20loaders%20%7D)%20%3D%3E%20loaders.Cart.cart()%5Cn%20%20%20%20.map((cart)%20%3D%3E%20cart%20%7C%7C%20%7B%20items_qty%3A%200%20%7D)%5Cn%20%20%20%20.toPromise()%5Cn%7D%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22doc%22%3Afalse%7D%7D) I found while _prettifying_ our project.

Please let me know if you need more details on this

**Comment by vjeux** [comment]
@real34 this is an awesome list, it's definitely going in as a test!

**Comment by vjeux** [comment]
Just pushed an update which fixes all the tests. Thanks a lot for this list!

This is how it now looks: https://gist.github.com/vjeux/83488a218252b4030daf6b0d01527fb6

**Comment by SimenB** [comment]
Amazing, can't wait for this! This is the only thing prettier has done that reduced readability for me, maybe except for #1099. Good job!

**Comment by real34** [comment]
I agree with @SimenB, the result is awesome! Thank you for your time on this.

The only one I find weird in the test is: https://gist.github.com/vjeux/83488a218252b4030daf6b0d01527fb6#file-test-js-L22

``` diff
const loadPricesBatch = skus => {
    const params = {};
-   return Observable.fromPromise(
-     axiosInstance.get("/frontcommerce/price", { params }),
-   )
+   return Observable
+     .fromPromise(
+       axiosInstance.get("/frontcommerce/price", { params }),
+     )
      .map(response => response.data)
      .map(reorderForSkus(skus));
};
```

... but it is a matter of taste I guess :)

**Comment by vjeux** [comment]
For fromPromise, if the call fit in a single line it would have stayed on the same line as Observable. It's a bit weird here but there were other cases where it completely broke so I figured it would be a good tradeoff

File: src/printer.js [hunk 2]
```diff
@@ -2106,11 +2122,29 @@ function printArgumentsList(path, options, print) {
...
       ? printed.slice(1).some(willBreak)
       : printed.slice(0, -1).some(willBreak);
+
+    let printedLastArgExpanded;
+    if (shouldGroupFirst) {
+      printedLastArgExpanded = printed;
+    } else {
+      // We want to print the last argument with a special flag
+      let i = 0;
+      path.each(function(argPath) {
```
Comment (jlongster) [inline src/printer.js]: What is this `path.each` for? `argPath` is never even used

File: src/printer.js [hunk 3]
```diff
@@ -2106,11 +2122,29 @@ function printArgumentsList(path, options, print) {
...
+        if (i++ === args.length - 1) {
+          printedLastArgExpanded = printed
+            .slice(0, -1)
+            .concat(path.call(
+              print.withArgs({expandLastArg: true})
```
Comment (jlongster) [inline src/printer.js]: I don't understand why you need `withArgs`. Why not do `path.call(p => print(p, { expandLastArg: true })))`? Am I missing what's wrong there?

**Comment by jlongster** [comment]
Two small questions, but this looks good in general!

File: src/printer.js (hunk 3 above)
Comment (vjeux) [inline src/printer.js]: The second argument `options` are the printer options (printWidth, useTabs...) and not calling the function with specific arguments.

I thought about doing `{...options, expandLastArg: true}` but the issue is that this option is forwarded down recursively. So this would mean that I'd also have to make sure to unset the expandLastArg before forwarding down otherwise the entire sub-tree would have this flag set, instead of just the node.

File: src/printer.js (hunk 2 above)
Comment (vjeux) [inline src/printer.js]: `path.call` below should instead be `argPath.call`. It happens to be working because `path` is being mutated during the traversal. Here I want to iterate on all the arguments and reprint the last one with the special flag.

File: src/printer.js (hunk 3 above)
Comment (jlongster) [inline src/printer.js]: `print` is not `genericPrint`. This is pretty confusing, and we should probably make this clearer somehow. `print` is the function passed in, which is `printGenerically`, which you adds the `args` argument below. The function in turn calls `genericPrint` with all this arguments: `p, options, printGenerically, args`. So the options that I passed in my snippet is the `args` that is being passed as the fourth argument to `genericPrint`.

I think that's right?

Comment (vjeux) [inline src/printer.js]: It works, you are totally right! Just updated the PR :)

**Comment by jlongster** [comment]
Awesome, looks good to me!

File: src/printer.js (hunk 3 above)
Comment (jlongster) [inline src/printer.js]: Cool! Yeah, how those functions work are super confusing, and we probably don't need it to be that way (it was from recast which did a lot of other things where this was more helpful)
